### PR TITLE
util::key add serde de/serialization

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

**Comment by elichai** [comment]
Maybe put the serde_test in `dev-dependencies`?
So it won't be compiled unless it's running the tests

**Comment by apoelstra** [comment]
Well, `dev-dependencies` can't be optional. So there's a tradeoff there..do we want to force all testers to compile it, or all serde users to compile it?

**Comment by elichai** [comment]
Ha, didn't know that it wasn't possible.
I would argue that worse testing times is preferable over production library size and compile time when users use this library.

**Comment by apoelstra** [comment]
cc @TheBlueMatt what do you think about this decision?

File: src/util/key.rs [hunk 1]
```diff
@@ -197,6 +198,118 @@ impl ops::Index<ops::RangeFull> for PrivateKey {
     }
 }
 
+#[cfg(feature = "serde")]
+impl ::serde::Serialize for PrivateKey {
+    fn serialize<S: ::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        s.collect_str(self)
```
Comment (stevenroose) [inline src/util/key.rs]: Hmm. I didn't know about this method. Would be nice if there was an equivalent for just using `FromStr` in the Deserialize.

Comment (apoelstra) [reply src/util/key.rs]: Well, unfortunately `FromStr` allocates..

File: src/util/key.rs [hunk 2]
```diff
@@ -197,6 +198,118 @@ impl ops::Index<ops::RangeFull> for PrivateKey {
...
+            }
+
+            fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+            where
+                E: ::serde::de::Error,
+            {
+                if let Ok(s) = ::std::str::from_utf8(v) {
+                    PrivateKey::from_str(s).map_err(E::custom)
+                } else {
+                    return Err(E::invalid_value(::serde::de::Unexpected::Bytes(v), &self));
```
Comment (sgeisler) [inline src/util/key.rs]: unneeded `return`

File: src/util/key.rs [hunk 3]
```diff
@@ -197,6 +198,118 @@ impl ops::Index<ops::RangeFull> for PrivateKey {
...
+                }
+
+                fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+                where
+                    E: ::serde::de::Error,
+                {
+                    if let Ok(hex) = ::std::str::from_utf8(v) {
+                        PublicKey::from_str(hex).map_err(E::custom)
+                    } else {
+                        return Err(E::invalid_value(::serde::de::Unexpected::Bytes(v), &self));
```
Comment (sgeisler) [inline src/util/key.rs]: unneeded `return`

**Comment by apoelstra** [comment]
I don't understand how any property of the serializer implies anything about the deserializer.

As for what methods I chose to implement, I copied them from the existing serde impl for `String` mostly.

File: src/util/key.rs [hunk 4]
```diff
@@ -197,6 +198,118 @@ impl ops::Index<ops::RangeFull> for PrivateKey {
...
+
+        d.deserialize_str(WifVisitor)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl ::serde::Serialize for PublicKey {
+    fn serialize<S: ::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        if s.is_human_readable() {
+            s.collect_str(&self.to_string())
```
Comment (stevenroose) [inline src/util/key.rs]: I think the to_string here is redundant, collect_str uses Display as well

Comment (apoelstra) [reply src/util/key.rs]: You're right, not sure what I was thinking there.

**Pull request merged**